src/compartment.c

```diff
@@ -569,113 +710,307 @@ comp_clean(struct Compartment* to_clean)
     close(to_clean->fd);
     if (to_clean->mapped)
     {
-        for (size_t i = 0; i < to_clean->seg_count; ++i)
-        {
-            free(to_clean->segs[i]);
-            // TODO unmap
-        }
+        // TODO unmap
     }
     else if (to_clean->mapped_full)
     {
         // TODO unmap
     }
+
+    for (size_t i = 0; i < to_clean->seg_count; ++i)
+    {
+        free(to_clean->segs[i]);
+    }
+    free(to_clean->segs);
+
     for (size_t i = 0; i < to_clean->entry_point_count; ++i)
     {
         free((char*) to_clean->comp_fns[i]->fn_name);
         free(to_clean->comp_fns[i]);
     }
+
+    for (size_t i = 0; i < to_clean->rela_maps_count; ++i)
+    {
+        free(to_clean->rela_maps[i].rela_name);
+    }
+    free(to_clean->rela_maps);
+
+    struct LibDependency* ld;
+    for (size_t i = 0; i < to_clean->lib_deps_count; ++i)
+    {
+        ld = to_clean->lib_deps[i];
+        free(ld->lib_name);
+        free(ld->lib_path);
+        for (size_t j = 0; j < ld->lib_segs_count; ++j)
+        {
+            free(ld->lib_segs[j]);
+        }
+        free(ld->lib_segs);
+        for (size_t j = 0; j < ld->lib_syms_count; ++j)
+        {
+            free(ld->lib_syms[j].sym_name);
+        }
+        free(ld->lib_syms);
+        free(ld);
+    }
+    free(to_clean->lib_deps);
+    free(to_clean->intercept_patches);
+
+
     free(to_clean);
     // TODO
 }
 
-void
-log_new_comp(struct Compartment* to_log)
+/*******************************************************************************
+ * Helper functions
+ ******************************************************************************/
+
+static
+ssize_t
+do_pread(int fd, void* buf, size_t count, off_t offset)
 {
-    comps = realloc(comps, sizeof(comps) + sizeof(struct Compartment));
-    comps[to_log->id] = to_log;
+    if(errno != 0)
```

Comment:
I think this function should document that `errno` should come from a certain function and/or be set to zero explicitly?

Reply:
`errno` isn't nebulous -- but it is evil! Some libc functions set `errno` and they have to explicitly document that they do so. But not all functions do so. To add insult to injury, some libc functions set errno in some situations and not others, so you have to be *really* careful about when you can check errno at all! [Yes this latter thing surprised me when I found out about it!]

Summary: we can't randomly check `errno`. You have to know a) which function call immediately preceeds the check for `errno` b) what the conditions are that the function will set `errno`.

Reply:
Is the ideal case not to check for `errno` whenever a function that sets it is called? As such, if I understand this point correctly, I should remove this pre-check in this function, and also go through the code and find all these functions?

Reply:
> if I understand this point correctly, I should remove this pre-check in this function, and also go through the code and find all these functions?

I'm afraid you're correct.

As this suggests, `errno` is a real mess. One has also to bear in mind also that some functions only set `errno` in some situations. So sometimes you have to check the return code of the function and only then can you safely check (or not...) `errno`.

Reply:
After reading in more detail about `errno` (I found a nice page with lots of examples [^1]), it makes complete sense to remove the preamble check - done in https://github.com/capablevms/CHERI-ELF-comp/pull/13/commits/eaad0b8b6953b34c18f805132928d04bc06db9ca.

However, I think going through each function call that might fail and set `errno` that I haven't guarded against might be a bit too much, considering I expect the code to still be in flux. I can't think of a way to search "give me all functions that might set `errno`" to do this, so I guess it'd have to be manual. Besides, I know guards are lacking, such as no guards against most `malloc` calls.

I would suggest I keep this in mind for future code / rewrites where I might use `errno`-changing functions, and leave the existing code as is otherwise.